File: lib/domain/field_cultivation/gateways/field_cultivation_gateway.rb
# frozen_string_literal: true

module Domain
  module FieldCultivation
    module Gateways
      class FieldCultivationGateway
        # @return [Domain::FieldCultivation::Dtos::FieldCultivationPlanAccessSnapshot]
        # @raise [Domain::Shared::Exceptions::RecordNotFound]
        def find_plan_access_snapshot_by_field_cultivation_id(field_cultivation_id)
          raise NotImplementedError,
                "Subclasses must implement find_plan_access_snapshot_by_field_cultivation_id"
        end

        # @return [Domain::FieldCultivation::Dtos::FieldCultivationApiSummary]
        # @raise [Domain::Shared::Exceptions::RecordNotFound]
        def find_api_summary_by_field_cultivation_id(field_cultivation_id)
          raise NotImplementedError,
                "Subclasses must implement find_api_summary_by_field_cultivation_id"
        end

        # @return [Domain::FieldCultivation::Dtos::FieldCultivationApiUpdateOutput]
        def update_field_cultivation_schedule(field_cultivation_id:, start_date:, completion_date:, cultivation_days: nil)
          raise NotImplementedError, "Subclasses must implement update_field_cultivation_schedule"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field_cultivation/gateways/field_cultivation_gateway.rs
use crate::field_cultivation::dtos::FieldCultivationApiUpdateOutput;
use crate::field_cultivation::dtos::{
    FieldCultivationApiSummary, FieldCultivationPlanAccessSnapshot,
};

/// Ruby: `Domain::FieldCultivation::Gateways::FieldCultivationGateway`
pub trait FieldCultivationGateway: Send + Sync {
    fn find_plan_access_snapshot_by_field_cultivation_id(
        &self,
        field_cultivation_id: i64,
    ) -> Result<FieldCultivationPlanAccessSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn find_api_summary_by_field_cultivation_id(
        &self,
        field_cultivation_id: i64,
    ) -> Result<FieldCultivationApiSummary, Box<dyn std::error::Error + Send + Sync>>;

    fn update_field_cultivation_schedule(
        &self,
        field_cultivation_id: i64,
        start_date: &str,
        completion_date: &str,
        cultivation_days: Option<i32>,
    ) -> Result<FieldCultivationApiUpdateOutput, Box<dyn std::error::Error + Send + Sync>>;
}
